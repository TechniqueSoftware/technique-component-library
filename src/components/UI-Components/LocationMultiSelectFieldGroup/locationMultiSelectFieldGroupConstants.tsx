export const REGIONS_GROUP_ID = 'regions';
export const LOCATIONS_GROUP_ID = 'locations';
