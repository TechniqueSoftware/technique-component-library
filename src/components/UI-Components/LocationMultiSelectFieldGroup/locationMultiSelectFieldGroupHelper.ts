import { MultiSelectOptionData } from '../MultiSelectFieldGroup/MultiSelectSearchList/MultiSelectSearchList';
import pluralize from 'pluralize';

export interface RenderButtonLabelParams {
  regionList: MultiSelectOptionData[];
  clubLocationList: MultiSelectOptionData[];
  regionValues: Set<string>;
  clubLocationValues: Set<string>;
}

const MAX_BUTTON_LABEL_LENGTH = 25;
const ALL_LOCATIONS_LABEL = 'All Locations';

function truncateLabel(label: string) {
  return label.length > MAX_BUTTON_LABEL_LENGTH ? `${label.substring(0, MAX_BUTTON_LABEL_LENGTH)}...` : label;
}

function getLabelStringForMatchingOptionData(value: string, optionData: MultiSelectOptionData[]) {
  const option = optionData.find((region: MultiSelectOptionData) => region.value === value);
  return truncateLabel(option.labelString);
}

function regionPlural(count: number) {
  return pluralize('Region', count);
}

function locationPlural(count: number) {
  return pluralize('Location', count);
}

function regionsCountLabel(count: number) {
  return `${count} ${regionPlural(count)}`;
}

function locationsCountLabel(count: number) {
  return `${count} ${locationPlural(count)}`;
}

/**
 * Gets the appropriate label for our button based on what values are selected.
 *
 * There is now a more generic version of this for general use. See the `getButtonLabel` method in multiSelectFieldGroupButtonHelper.ts
 * @param params
 */
export function getRegionsOrLocationsSelectedLabel(params: RenderButtonLabelParams) {
  const {
    regionValues,
    clubLocationValues,
    regionList,
    clubLocationList,
  } = params;

  let label = ALL_LOCATIONS_LABEL;

  const locationCount = clubLocationValues.size;
  const regionCount = regionValues.size;

  if (regionCount === 1 && locationCount === 0) {
    const value = regionValues.values().next().value;
    label = getLabelStringForMatchingOptionData(value, regionList);
  } else if (locationCount === 1 && regionCount === 0) {
    const value = clubLocationValues.values().next().value;
    label = getLabelStringForMatchingOptionData(value, clubLocationList);
  } else if (locationCount > 1 && regionCount === 0) {
    label = locationsCountLabel(locationCount);
  } else if (regionCount > 1 && locationCount === 0) {
    label = regionsCountLabel(regionCount);
  } else if (regionCount > 0 && locationCount > 0) {
    label = `${regionsCountLabel(regionCount)}, ${locationsCountLabel(locationCount)}`;
  }
  return label;
}
