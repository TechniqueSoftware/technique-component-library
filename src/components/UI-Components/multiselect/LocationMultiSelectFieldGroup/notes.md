# LocationMultiSelectFieldGroup
This component wraps the generic `MultiSelectFieldGroup` component and passes in location/region specific labels and 
helper text. The component provides features to allow searching as well as an option to indicate that they want to 
select all locations/regions including any locations/regions that would be adding in the future.


## Usage
This component was created initially to be used in the email and SMS campaign creation views as well as the email
template views.

## Issues

### Performance
There are known performance issues when handling hundreds or thousands of items in the lists. Additional performance
optimizations could be made but we've decided to wait until we determine its necessary.

## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/CR-579
