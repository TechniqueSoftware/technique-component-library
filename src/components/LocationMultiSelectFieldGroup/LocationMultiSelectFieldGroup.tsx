import * as React from 'react';
import { pluralize } from '../../services/pluralizeService';
import { LOCATIONS_GROUP_ID, REGIONS_GROUP_ID } from './locationMultiSelectFieldGroupConstants';
// noinspection TypeScriptPreferShortImport
import {
  MultiSelectGroupValueWrapper,
  MultiSelectOptionData
} from '../MultiSelectFieldGroup/MultiSelectSearchList/MultiSelectSearchList';
import MultiSelectFieldGroup, { MultiSelectFieldGroupProps } from '../MultiSelectFieldGroup/MultiSelectFieldGroup';

export interface LocationMultiSelectFieldGroupProps {
  fieldSetLabelText?: string | React.ReactNode;
  selectLabelComponent?: React.ReactElement<any> | undefined;
  readOnly?: boolean;
  helperText?: string;
  error?: boolean;
  showRadioGroup?: boolean;
  showFieldSetLabelText?: boolean;
  showSelectedLocationHelperText?: boolean;
  selectAll?: string | 'true' | 'false';
  listContainerHeight?: number;
  regionList?: MultiSelectOptionData[];
  regionValues?: Set<string>;
  clubLocationList?: MultiSelectOptionData[];
  clubLocationValues?: Set<string>;
  applyButton?: boolean;
  clearButton?: boolean;
  onChange?: (changes: { regions: Set<any>, locations: Set<any>, selectAll: boolean}) => void;
  onCheckboxGroupChange?: (values: { regions: Set<any>, locations: Set<any> }) => void;
  onAllOrSelectRadioGroupChange?: (event: React.ChangeEvent<{}>, value: string) => void;
}

export class LocationMultiSelectFieldGroup extends React.PureComponent<LocationMultiSelectFieldGroupProps> {

  static defaultProps: Partial<LocationMultiSelectFieldGroupProps> = {
    selectAll: 'false',
    fieldSetLabelText: '',
    helperText: '',
    readOnly: false,
    regionList: [],
    clubLocationList: [],
    regionValues: new Set(),
    clubLocationValues: new Set(),
    listContainerHeight: 450,
    showRadioGroup: false,
    showFieldSetLabelText: false,
    showSelectedLocationHelperText: true
  };

  renderSelectLabelComponent() {
    const { selectLabelComponent } = this.props;
    if (selectLabelComponent) {
      return React.cloneElement(selectLabelComponent);
    }
    return 'Select Locations';
  }

  handleCheckboxGroupChange = (groupValues: MultiSelectGroupValueWrapper[]) => {
    const {
      selectAll,
      onChange,
      onCheckboxGroupChange,
    } = this.props;

    const regions = groupValues.find((item) => item.groupId === REGIONS_GROUP_ID);
    const locations = groupValues.find((item) => item.groupId === LOCATIONS_GROUP_ID);
    const regionValues = regions ? regions.values : new Set();
    const locationValues = locations ? locations.values : new Set();

    if (onChange) onChange({ regions: regionValues, locations: locationValues, selectAll: selectAll === 'true' });
    if (onCheckboxGroupChange) onCheckboxGroupChange({ regions: regionValues, locations: locationValues });
  }

  handleSelectAllChange = (event: React.ChangeEvent<{}>, value: string) => {
    const {
      onChange,
      regionValues,
      clubLocationValues,
      onAllOrSelectRadioGroupChange,
    } = this.props;

    if (onChange) onChange({ regions: regionValues, locations: clubLocationValues, selectAll: value === 'true' });
    if (onAllOrSelectRadioGroupChange) onAllOrSelectRadioGroupChange(event, value);
  }

  renderHelperText() {
    const {
      helperText,
      regionValues,
      clubLocationValues,
      showSelectedLocationHelperText
    } = this.props;

    if (helperText) return helperText;

    if (!showSelectedLocationHelperText) return null;

    const regionsCount = regionValues.size;
    const locationCount = clubLocationValues.size;
    const regionsAndLocationsSelected = regionsCount > 0 && locationCount > 0;
    const regionsOrLocationSelected = regionsCount > 0 || locationCount > 0;

    let message = '';

    if (regionsCount > 0) {
      message = `${regionsCount} ${pluralize('region', regionsCount)}`;

    }
    if (regionsAndLocationsSelected) {
      message += ' and ';
    }

    if (locationCount > 0) {
      message += `${locationCount} ${pluralize('location', locationCount)}`;
    }

    if (regionsOrLocationSelected) {
      message += ' selected';
    }

    return message;
  }

  render() {
    const {
      clubLocationList,
      regionList,
      selectAll,
      regionValues,
      clubLocationValues,
      showRadioGroup,
      fieldSetLabelText,
      readOnly,
      error,
      showFieldSetLabelText,
      showSelectedLocationHelperText,
      listContainerHeight,
    } = this.props;

    const radioInputName = 'selectAllActiveAndFutureRegions';

    const props: MultiSelectFieldGroupProps = {
      listContainerHeight,
      showFieldSetLabelText,
      readOnly,
      error,
      fieldSetLabelText,
      showRadioGroup,
      showHelperText: showSelectedLocationHelperText,
      helperText: this.renderHelperText(),
      onCheckboxGroupChange: this.handleCheckboxGroupChange,
      groupValueWrapperList: [
        {
          groupId: REGIONS_GROUP_ID,
          values: regionValues
        }, {
          groupId: LOCATIONS_GROUP_ID,
          values: clubLocationValues
        }],
      radioGroupProps: {
        label: 'Accessible to locations',
        name: radioInputName,
        value: selectAll
      },
      onRadioGroupChange: this.handleSelectAllChange,
      selectAllRadioInputProps: {
        label: `All Active and Future Locations`,
        value: 'true',
        name: radioInputName,
      },
      selectSomeRadioInputProps: {
        name: radioInputName,
        label: this.renderSelectLabelComponent(),
        value: 'false',
      },
      multiSelectOptionGroupDataList: [
        {
          groupLabel: 'Regions',
          groupId: REGIONS_GROUP_ID,
          multiSelectOptionDataList: regionList
        },
        {
          groupLabel: 'Locations',
          groupId: LOCATIONS_GROUP_ID,
          multiSelectOptionDataList: clubLocationList
        }
      ]
    };
    return (<MultiSelectFieldGroup {...props} />);
  }
}

export default LocationMultiSelectFieldGroup;
