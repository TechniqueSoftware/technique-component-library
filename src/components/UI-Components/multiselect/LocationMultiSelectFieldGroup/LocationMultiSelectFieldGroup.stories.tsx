import * as React from 'react';
import { createCUICStoriesOf } from '../../utils/utils';
import { MODULE_NAME } from '../../constants';
import {
  MultiSelectGroupValueWrapper,
  MultiSelectOptionData
} from '../../MultiSelectFieldGroup/MultiSelectSearchList/MultiSelectSearchList';
import clubOsTheme from '../../storyThemes';
import { IconTooltip, IconTooltipProps } from '../../Tooltip/IconTooltip/IconTooltip';
import { LocationMultiSelectFieldGroupProps } from
    '../../LocationMultiSelectFieldGroup/LocationMultiSelectFieldGroup';
import { LocationMultiSelectFieldGroup } from '../../LocationMultiSelectFieldGroup/LocationMultiSelectFieldGroup';
import { boolean, number, text } from '@storybook/addon-knobs';
import { createRandomCompanyNameWordsArray } from '../../utils/fakeData';
import { ThemeProvider } from '@material-ui/styles';
import HelpRounded from '@material-ui/icons/HelpRounded';

const clubLocationNames = createRandomCompanyNameWordsArray(2000);
const regionLocationNames = createRandomCompanyNameWordsArray(2000);

interface LocationMultiSelectStoryComponentState {
  selectAll: string | 'true' | 'false';
  selectedRegions: Set<string>;
  selectedLocations: Set<string>;
  groupValueWrapper: MultiSelectGroupValueWrapper[];
}

interface LocationMultiSelectStoryComponentProps {
  showSelectedLocationHelperText: boolean;
  showFieldSetLabelText: boolean;
  helperText: string;
  fieldSetLabelText: string;
  readOnly: boolean;
  locationItemCount: number;
  regionItemCount: number;
  showRadioGroup: boolean;
  error: boolean;
  maxWidth: number;
}

class LocationMultiSelectStoryComponent extends React.Component<LocationMultiSelectStoryComponentProps,
  LocationMultiSelectStoryComponentState> {

  static defaultProps = {
    readOnly: false
  };

  constructor(props: any) {
    super(props);
    this.state = {
      selectAll: 'false',
      selectedRegions: new Set(),
      selectedLocations: new Set(),
      groupValueWrapper: [],
    };

    this.onSelectAllRadioGroupChange = this.onSelectAllRadioGroupChange.bind(this);
    this.handleCheckboxGroupChange = this.handleCheckboxGroupChange.bind(this);
  }

  onSelectAllRadioGroupChange(event: any, value: string) {
    this.setState({ selectAll: value });
  }

  handleCheckboxGroupChange(values: { regions: Set<string>, locations: Set<string> }) {
    this.setState({ selectedRegions: values.regions, selectedLocations: values.locations });
  }

  static createListItem(label: string, index: number): MultiSelectOptionData {
    let smsComponent;
    let labelString = label;
    const textCampaignsDisabled = '(text campaigns disabled)';
    if ((index % 3) === 0) {
      labelString = `${label} {textCampaignsDisabled}`;
      smsComponent = (<span>{label} <i>{textCampaignsDisabled}</i></span>);
    }
    return {
      labelString,
      value: label,
      labelComponent: smsComponent,
      name: 'mockName1',
      disabled: !!smsComponent
    };
  }

  createListItems(names: string[], count: number): MultiSelectOptionData[] {
    return Array(count).fill('').map((value, index) => {
      return LocationMultiSelectStoryComponent.createListItem(names[index], index);
    });
  }

  render() {
    const {
      locationItemCount,
      regionItemCount,
      showRadioGroup,
      helperText,
      error,
      maxWidth,
      readOnly,
      showFieldSetLabelText,
      fieldSetLabelText,
      showSelectedLocationHelperText
    } = this.props;
    const {
      selectAll,
      selectedLocations,
      selectedRegions
    } = this.state;

    const iconTooltipProps: IconTooltipProps = {
      iconProps: {
        component: HelpRounded
      },
      tooltipText: 'Locations that do not have bulk texting enabled cannot be selected. Contact support@club-os.com for help.'
    };

    const props: LocationMultiSelectFieldGroupProps = {
      showSelectedLocationHelperText,
      showFieldSetLabelText,
      helperText,
      error,
      readOnly,
      selectAll,
      showRadioGroup,
      fieldSetLabelText,
      clubLocationList: this.createListItems(clubLocationNames, locationItemCount),
      selectLabelComponent: <span>Select Locations <IconTooltip {...iconTooltipProps} /></span>,
      onCheckboxGroupChange: this.handleCheckboxGroupChange,
      regionValues: selectedRegions,
      clubLocationValues: selectedLocations,
      onAllOrSelectRadioGroupChange: this.onSelectAllRadioGroupChange,
      onChange: changes => console.log('LocationMultiSelectFieldGroup onChange', changes),
    };

    if (regionItemCount > 0) {
      props.regionList = this.createListItems(regionLocationNames, regionItemCount);
    }

    return (
      <ThemeProvider theme={clubOsTheme}>
        <div style={{ maxWidth }}>
          <LocationMultiSelectFieldGroup {...props} />
        </div>
      </ThemeProvider>);
  }
}

function renderSelectLocationsComponent() {

  const props: LocationMultiSelectStoryComponentProps = {
    showSelectedLocationHelperText: boolean('Show selected location helper text', true),
    helperText: text('Helper Text', ''),
    fieldSetLabelText: text('Fieldset label text', 'Make template accessible to'),
    showFieldSetLabelText: boolean('Show fieldset label text', true),
    showRadioGroup: boolean('Show radio group', true),
    readOnly: boolean('Read only', false),
    error: boolean('Error', false),
    regionItemCount: number('Region count (click a checkbox to force an update)', 10),
    locationItemCount: number('Location count (click a checkbox to force an update)', 200),
    maxWidth: number('Container max width', 500)
  };
  return <LocationMultiSelectStoryComponent {...props} />;
}

createCUICStoriesOf(MODULE_NAME.MULTISELECT, module)
  .addCUICStory({
    render: renderSelectLocationsComponent,
    name: 'LocationMultiSelectFieldGroup',
    notes: require('./notes.md').default
  });
