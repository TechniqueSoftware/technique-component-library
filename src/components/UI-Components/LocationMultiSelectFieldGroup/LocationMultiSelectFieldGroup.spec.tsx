import { LocationMultiSelectFieldGroup, LocationMultiSelectFieldGroupProps } from './LocationMultiSelectFieldGroup';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('LocationMultiSelectFieldGroup', () => {

  const baseProps: LocationMultiSelectFieldGroupProps = {
    selectLabelComponent: undefined,
    showRadioGroup: false,
    regionList: [],
    clubLocationList: [],
    clubLocationValues: new Set(),
    regionValues: new Set(),
    selectAll: 'true',
    onChange: jest.fn(),
    onCheckboxGroupChange: jest.fn(),
    onAllOrSelectRadioGroupChange: jest.fn(),
  };

  test('should render with minimal props', () => {
    const props = { ...baseProps };
    const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  test('should pass on props matching snapshot', () => {
    const props = { ...baseProps };
    const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
    const multiSelectFieldGroupWrapper = shallowWrapper.find('WithStyles(MultiSelectFieldGroup)');
    expect(multiSelectFieldGroupWrapper.props()).toMatchSnapshot();
  });

  describe('helperText', () => {
    test('should be blank when no locations or regions selected', () => {
      const props = { ...baseProps };
      const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
      const multiSelectFieldGroupWrapper = shallowWrapper.find('WithStyles(MultiSelectFieldGroup)');
      const helperText = multiSelectFieldGroupWrapper.prop('helperText');
      expect(helperText).toEqual('');
    });

    test('should include "region" when a single region is provided', () => {
      const props: LocationMultiSelectFieldGroupProps = {
        ...baseProps,
        regionValues: new Set(['1'])
      };
      const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
      const multiSelectFieldGroupWrapper = shallowWrapper.find('WithStyles(MultiSelectFieldGroup)');
      const helperText = multiSelectFieldGroupWrapper.prop('helperText');
      expect(helperText).toEqual('1 region selected');
    });

    test('should include "regions" when a multiple regions are provided', () => {
      const props: LocationMultiSelectFieldGroupProps = {
        ...baseProps,
        regionValues: new Set(['1', '2', '3'])
      };
      const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
      const multiSelectFieldGroupWrapper = shallowWrapper.find('WithStyles(MultiSelectFieldGroup)');
      const helperText = multiSelectFieldGroupWrapper.prop('helperText');
      expect(helperText).toEqual('3 regions selected');
    });

    test('should include "locations" and "regions" when a multiple regions are provided', () => {
      const props: LocationMultiSelectFieldGroupProps = {
        ...baseProps,
        regionValues: new Set(['1', '2', '3']),
        clubLocationValues: new Set(['1', '2', '3'])
      };
      const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
      const multiSelectFieldGroupWrapper = shallowWrapper.find('WithStyles(MultiSelectFieldGroup)');
      const helperText = multiSelectFieldGroupWrapper.prop('helperText');
      expect(helperText).toEqual('3 regions and 3 locations selected');
    });
  });

  test('should show the helper text when no "locations" or "regions" are selected', () => {
    const props: LocationMultiSelectFieldGroupProps = {
      ...baseProps,
      helperText: 'Required',
    };
    const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
    const multiSelectFieldGroupWrapper = shallowWrapper.find('WithStyles(MultiSelectFieldGroup)');
    const helperText = multiSelectFieldGroupWrapper.prop('helperText');
    expect(helperText).toEqual('Required');
  });

  test('should show the helper text when "locations" and "regions" are selected', () => {
    const props: LocationMultiSelectFieldGroupProps = {
      ...baseProps,
      helperText: 'Check your selections',
      regionValues: new Set(['1', '2', '3']),
      clubLocationValues: new Set(['1', '2', '3']),
    };
    const shallowWrapper = shallow(<LocationMultiSelectFieldGroup {...props} />);
    const multiSelectFieldGroupWrapper = shallowWrapper.find('WithStyles(MultiSelectFieldGroup)');
    const helperText = multiSelectFieldGroupWrapper.prop('helperText');
    expect(helperText).toEqual('Check your selections');
  });

});
