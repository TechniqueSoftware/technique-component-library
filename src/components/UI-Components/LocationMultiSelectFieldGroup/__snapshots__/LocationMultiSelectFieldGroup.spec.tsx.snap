// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationMultiSelectFieldGroup should pass on props matching snapshot 1`] = `
Object {
  "error": undefined,
  "fieldSetLabelText": "",
  "groupValueWrapperList": Array [
    Object {
      "groupId": "regions",
      "values": Set {},
    },
    Object {
      "groupId": "locations",
      "values": Set {},
    },
  ],
  "helperText": "",
  "listContainerHeight": 450,
  "multiSelectOptionGroupDataList": Array [
    Object {
      "groupId": "regions",
      "groupLabel": "Regions",
      "multiSelectOptionDataList": Array [],
    },
    Object {
      "groupId": "locations",
      "groupLabel": "Locations",
      "multiSelectOptionDataList": Array [],
    },
  ],
  "onCheckboxGroupChange": [Function],
  "onRadioGroupChange": [Function],
  "radioGroupProps": Object {
    "label": "Accessible to locations",
    "name": "selectAllActiveAndFutureRegions",
    "value": "true",
  },
  "readOnly": false,
  "selectAllRadioInputProps": Object {
    "label": "All Active and Future Locations",
    "name": "selectAllActiveAndFutureRegions",
    "value": "true",
  },
  "selectSomeRadioInputProps": Object {
    "label": "Select Locations",
    "name": "selectAllActiveAndFutureRegions",
    "value": "false",
  },
  "showFieldSetLabelText": false,
  "showHelperText": true,
  "showRadioGroup": false,
}
`;

exports[`LocationMultiSelectFieldGroup should render with minimal props 1`] = `"<WithStyles(MultiSelectFieldGroup) listContainerHeight={450} showFieldSetLabelText={false} readOnly={false} error={[undefined]} fieldSetLabelText=\\"\\" showRadioGroup={false} showHelperText={true} helperText=\\"\\" onCheckboxGroupChange={[Function (anonymous)]} groupValueWrapperList={{...}} radioGroupProps={{...}} onRadioGroupChange={[Function (anonymous)]} selectAllRadioInputProps={{...}} selectSomeRadioInputProps={{...}} multiSelectOptionGroupDataList={{...}} />"`;
